### PR TITLE
Add midnight? predicate to check if a time is equal to midnight

### PR DESCRIPTION
**Pull request by kieranj**
Convenience method for checking if a given time is equal to midnight.

Short-hand for

```
time == time.midnight
```

### CODE REVIEW

File: activesupport/test/core_ext/date_time_ext_test.rb [hunk 1]
```diff
@@ -72,6 +72,11 @@ def test_seconds_since_midnight
     assert_equal 86399,DateTime.civil(2005,1,1,23,59,59).seconds_since_midnight
   end
 
+  def test_midnight?
+    assert_equal true, Time.now.beginning_of_day.midnight?
```
Comment (georgeclaghorn) [inline activesupport/test/core_ext/date_time_ext_test.rb]: Should use `assert` and `assert_not` to test boolean values:

```ruby
assert Time.now.beginning_of_day.midnight?
assert_not Time.now.beginning_of_day.advance(:minutes => 1).midnight?
```

**Comment by fxn** [comment]
For the record, this patch defines a "predicate", rather than a "boolean".

**Comment by kieranj** [comment]
@fxn yes your right. I've updated the changelog

**Comment by georgeclaghorn** [comment]
@kieranj: Could you please squash your commits?

**Comment by kieranj** [comment]
@georgeclaghorn commits squashed

**Comment by egilburg** [comment]
What's the use case for this? We could also add predicates to all the other timepoints (e.g. beginning_of_day etc) but unlike the travel method ("give me the beginning of day"), the predicate would only work for an exact time instance down to the nanosecond and thus you wouldn't get it unless you previously requested that exact object anyways.

**Comment by kieranj** [comment]
The use case for this would be when you want to check if an instance of time is still set to midnight. 

Eg. You may default a datetime to midnight and then later wish to check if this is still the case or branch on this

It's something I've had in requirements a couple of times so thought this short-hand method would be useful to others.

**Comment by maclover7** [comment]
Can you rebase? Also, it looks like the tests are failing on Travis CI.

**Comment by kieranj** [comment]
Rebased as requested

**Review by gmcgibbon — COMMENTED** [review]
Still relevant in Rails 6 (168a9728dd00069efe04fec4b59d29e3752fbfb3). If there's already a `#midnight` method, I see no reason in not having a corresponding predicate. 👍 The problem I see is that [its an alias of `beginning_of_day`](https://github.com/rails/rails/blob/3101a4136bd62787e252d2658eee23001036fa0f/activesupport/lib/active_support/core_ext/date/calculations.rb#L70). Should we predicate the original definition name instead?

**Pull request closed without merging**